File: uploads/web/rc.js
import * as blockdata from "./blockdata.js";

let msgs = new Map(); // msgid => [successCallback, failCallback]
let notifyListeners = new Map(); // channel => callback
let msgid = 0;
let socket = new WebSocket("ws://10.244.65.57:7648/");
window.theSocket = socket;
//socket.onopen = function(){alert("old onopen called");}
let openPromise = new Promise(function(success, fail) {
  let onOpen = function() {
    console.log("Websocket open!");
    success(socket);
  }
  socket.onopen = onOpen;
  if (socket.readyState == WebSocket.OPEN) {
    onOpen();
  }
});

socket.onmessage = function(msgEv) {
  //window.debugData = msgEv.data;
  let res = new Response(msgEv.data);
  res.arrayBuffer().then(data => {
    let parsed = MessagePack.decode(data);
    if (parsed.ty == "results") {
      msgs.get(parsed.msgid)[0](parsed.rows);
      msgs.delete(parsed.msgid)
    } else if (parsed.ty == "prepared") {
      msgs.get(parsed.msgid)[0](parsed.id);
      msgs.delete(parsed.msgid)
    } else if (parsed.ty == "notification") {
      (notifyListeners.get(parsed.channel) || (function(){}))(parsed)
    } else if (parsed.ty == "error") {
      if (parsed.msgid) {
        msgs.get(parsed.msgid)[1](parsed.msg)
      } else {
        for (const [k,v] of msgs.entries()) {
          v[1](parsed.msg)
        }
      }
    } else {
      console.log(`ERR: unrecognized message type ${parsed.ty}`);
      socket.close()
    }
  });
}

function sqlQuery(str, params) {
  if(!params) {
    params = [];
  }
  console.log(str, params);
  let myMsgid = (msgid += 1);
  return openPromise.then(ws => ws.send(MessagePack.encode({ty: "query",  statement: str, params: params, msgid: myMsgid})))
    .then(_ => new Promise((success, fail) => msgs.set(myMsgid, [success, fail])));
}

let focusdiv = document.getElementById("focusdiv");
let statusdeets = document.getElementById("statusdeets");
let log = document.getElementById("log");
let channelEl = document.getElementById("channel");

notifyListeners.set("turtle_rc_feedback", function(parsed) {
    let payload = JSON.parse(parsed.payload);
    if(payload.channel != channelEl.value) return;
    let facingName = ["north", "east", "south", "west"][payload.position.facing];
    statusdeets.innerText = `fuel:${payload.fuel} x${payload.position.x} y${payload.position.y} z${payload.position.z} ${facingName}`;
    let logLine = `${payload.channel}: ${payload.cmd}: ${JSON.stringify(payload.res)}`;
    let logLineEl = document.createElement("pre");
    logLineEl.innerText = logLine;
    log.prepend(logLineEl);
});

sqlQuery("listen turtle_rc_feedback");
openPromise.then(function(socket) {
    let logLine = 'Socket open';
    let logLineEl = document.createElement("pre");
    logLineEl.innerText = logLine;
    log.prepend(logLineEl);
});

const keyCmd = {
    w: "forward",
    s: "back",
    a: "left",
    d: "right",
    q: "up",
    e: "down",
    r: "dig_up",
    f: "dig_fwd",
    v: "dig_down",
    t: "place_up",
    g: "place_fwd",
    b: "place_down"
};

focusdiv.addEventListener("keydown", function(ev) {
    if(ev.ctrlKey) return true;
    let cmd = keyCmd[ev.key];
    if(cmd == null) return true;
    ev.preventDefault();
    sqlQuery(
        "select pg_notify('turtle_rc_control', $1)::text",
        [{ty: "text", val: JSON.stringify({channel: channelEl.value, cmd: cmd})}]
    );
    return false;
});

File: uploads/rc.lua
require("paranoidLogger")("rc")
require("shellib")
local db = require("db"):default()
local tArgs = {
  ...
}
local channel = tArgs[1]
if channel == nil or channel == "" then
  error("bad channel")
end
getGlobalOffset()
local selectItem
selectItem = function()
  for i = 1, 16 do
    if not (turtle.getItemCount() == 0) then
      break
    end
    local sel = turtle.getSelectedSlot()
    sel = sel + 1
    if sel == 17 then
      sel = 1
    end
    turtle.select(sel)
  end
end
local control_channel = "turtle_rc_control"
local feedback_channel = "turtle_rc_feedback"
local process
process = function()
  return db:process()
end
local main
main = function()
  db:query("listen " .. control_channel)
  while true do
    local _continue_0 = false
    repeat
      local evName, id, parsed = os.pullEvent("database_notification")
      print(textutils.serialise(parsed))
      if id ~= db.id then
        _continue_0 = true
        break
      end
      if parsed.channel ~= control_channel then
        _continue_0 = true
        break
      end
      local payload = textutils.unserializeJSON(parsed.payload)
      print(textutils.serialise(payload))
      if payload.channel ~= channel then
        _continue_0 = true
        break
      end
      local res
      if payload.cmd == "up" then
        res = tryUp()
      else
        if payload.cmd == "down" then
          res = tryDown()
        else
          if payload.cmd == "forward" then
            res = tryForward()
          else
            if payload.cmd == "back" then
              res = tryBack()
            else
              if payload.cmd == "left" then
                turnLeft()
                res = true
              else
                if payload.cmd == "right" then
                  turnRight()
                  res = true
                else
                  if payload.cmd == "dig_up" then
                    res = {
                      turtle.digUp()
                    }
                  else
                    if payload.cmd == "dig_fwd" then
                      res = {
                        turtle.dig()
                      }
                    else
                      if payload.cmd == "dig_down" then
                        res = {
                          turtle.digDown()
                        }
                      else
                        if payload.cmd == "place_up" then
                          selectItem()
                          res = {
                            turtle.placeUp()
                          }
                        else
                          if payload.cmd == "place_fwd" then
                            selectItem()
                            res = {
                              turtle.place()
                            }
                          else
                            if payload.cmd == "place_down" then
                              selectItem()
                              res = {
                                turtle.placeDown()
                              }
                            else
                              res = "warn: unrecognized command"
                            end
                          end
                        end
                      end
                    end
                  end
                end
              end
            end
          end
        end
      end
      local inventory = { }
      for i = 1, 16 do
        inventory[i] = (turtle.getItemDetail(i) or {
          count = 0
        })
      end
      local response = textutils.serializeJSON({
        cmd = payload.cmd,
        position = globalPosition(),
        fuel = turtle.getFuelLevel(),
        inventory = inventory,
        res = res,
        channel = channel
      })
      db:query("select pg_notify($1, $2)::text", {
        ty = "text",
        val = feedback_channel
      }, {
        ty = "text",
        val = response
      })
      _continue_0 = true
    until true
    if not _continue_0 then
      break
    end
  end
end
return parallel.waitForAny(process, main)
